### PR TITLE
Start support for nameplate CD timer icons and code refactoring.

### PR DESCRIPTION
**Pull request by cont1nuity**
- Listening to "DBM_Timer*" events to show those as icons on the nameplate
- Timer icons support show, hide, pause, resume and update events, along with `keep`.
- Icons will now auto-hide if the timer runs out and `keep` is not set. This removes the need for `Schedule` hides, which could cause early hides if the icon/timer is updated.
- Support for CD swipe and timer text for existing nameplate icons
- Code refactoring to no longer use `texture` as global index, but `texture` only for existing nameplate icons and the `id` for the CD timer bars.
- Trying to keep the support for "handoff" to nameplate addons in place
- Both, Nameplate Icons and CD Timers, will be shown in the same line with a small space between them.

### CODE REVIEW

**Comment by cont1nuity** [comment]
Note: This is work in progress.

It appears to be generally working, but there is still refactoring to be done to remove all the existing looping and replace it with some better lookup for nameplates.
There is some general code cleanup to be done as well.

**Comment by QartemisT** [comment]
Current changed LGTM. As long as this is fully tested by Author, I'm happy with it.

**Review by QartemisT — APPROVED** [review]

File: DBM-Core/DBM-Nameplate.lua [hunk 1]
```diff
@@ -272,9 +272,14 @@ end
...
-local function SupportedNPMod()
+local function SupportedNPModIcons()
 	if not DBM.Options.UseNameplateHandoff then return false end
-	if _G["KuiNameplates"] or _G["TidyPlatesThreatDBM"] or _G["Plater"] then return true end
+	if _G["KuiNameplates"] or _G["TidyPlatesThreatDBM"] or (_G["Plater"] and Plater.db.profile.bossmod_support_enabled) then return true end
+	return false
+end
+local function SupportedNPModBars()
+	if not DBM.Options.UseNameplateHandoff then return false end
+	if _G["KuiNameplates"] or _G["TidyPlatesThreatDBM"] or (_G["Plater"] and Plater.db.profile.bossmod_support_bars_enabled) then return true end
```
Comment (MysticalOS) [inline DBM-Core/DBM-Nameplate.lua:282]: There is no need for Plater.db.profile.bossmod_support_bars_enabled check here. 
Unless the user has disabled handoff, if plater is installed it should use platers preference. I already added a login print that advices user to enable it in plater if it's off, or disable it in DBM. This current code would just disrespect platers setting and force plater users to use DBMs internal even though handoff is enabled. That's not good

Also only plater supports bars so shouldn't check kui or threatplates at all.

Comment (cont1nuity) [reply DBM-Core/DBM-Nameplate.lua]: Makes sense, thanks for the clarification.

**Comment by cont1nuity** [comment]
I am thinking about adding some option for direction of the icon row growth. Along with the offset, this would allow nice positioning, e.g. upper left corner, growing right.

File: DBM-Core/DBM-Nameplate.lua [hunk 2]
```diff
@@ -349,6 +412,336 @@ function nameplateFrame:Hide(isGUID, unit, spellId, texture, force, isHostile, i
...
+
+		if force then
+			if isFriendly then
+				DBM:FireEvent("BossMod_DisableFriendlyNameplates")
+			end
+			if isHostile then
+				DBM:FireEvent("BossMod_DisableHostileNameplates")
+			end
+		elseif unit then
+			DBM:FireEvent("BossMod_HideNameplateAura", isGUID, unit, texture)
```
Comment (Mitalie) [inline DBM-Core/DBM-Nameplate.lua:736]: This should pass `currentTexture` instead of `texture`

Comment (cont1nuity) [reply DBM-Core/DBM-Nameplate.lua]: Good catch, thank you. Fixed. :)

**Comment by cont1nuity** [comment]
> I am thinking about adding some option for direction of the icon row growth. Along with the offset, this would allow nice positioning, e.g. upper left corner, growing right.

This is implemented now, but option is not added to the menu: `DBM.Options.NPIconGrowthDirection` takes `"UP"`, `"DOWN"`, `"LEFT"`, `"RIGHT"` and `"CENTER"` (default if unset)

**Comment by MysticalOS** [comment]
yeah i've pretty much only used unitid for testing. it only even exists because that's way it was implemented in kui nameplates and bigwigs. it was more of a parity thing to support unitid and guid. but yeah i've yet to find a niche scenario where guid doesn't work for everything.

**Comment by cont1nuity** [comment]
`unitID` is usually a bit better performance wise in the base scenario, as you don't need to iterate all shown nameplates to find the right one. But that can easily be omitted via caching.

**Comment by MysticalOS** [comment]
I thought you didn't have to iterate anymore, you just use the get nameplate for unit token api call?

**Comment by cont1nuity** [comment]
From the documentation:
`Internally this function will iterate over a list of unit tokens to find a matching candidate. Unit tokens will be tested in the order given below, returning the first one that matches.`
Pretty sure this might be doing internal `UnitGUID` for those tokens it iterates...
Cache is safer :)

**Comment by MysticalOS** [comment]
It'd be nice if blizzard maintained a global cache so every addon doesn't need to do it.

**Comment by MysticalOS** [comment]
BTW, one thing I'd like to see is if Timer_Stop is called without a guid, it still kills the namplate icon.

The way DBM has typically worked that stop or cancel without args kills all variants of timer.

IE if for example we have timerobject:Stop(GUID) then it only stops timers matching that guid, or if you did timerobject:Stop(1, GUID) it'd match both 1 and guid args. But if you did timerobject:Stop() then it doesn't matter if it has a guid or not or a 1 or not, stop with no args is a hard stop on that object, period.

This eliminates a lot of extra work. For example right now i'm working on a council boss where if one member dies, it cancels ANOTHER bosses timer, and to do it with current plater callback expectations, I have to cache guids just so i can pass the expected guid to the other bosses stop timer. There are also many count timers in use that are currently requiring multiple stop calls again for same reason, because plater doesn't support hard stop. an example here: https://github.com/DeadlyBossMods/DBM-Dungeons/blob/master/DBM-Party-Dragonflight/BrackenhideHollow/HackclawsWarBand.lua#L215-L225 which is using stop all for DBM and stop with guid for plater. So this would need to be improved on both platers end and DBMs internal nameplate handling.

TL/DR, Timer_Stop should mean clear nameplate icon, guid or not. Now if this slows down having this ready for 10.2 I can use double stops for time being and it can be a separate post launch improvement PR down the road.

**Comment by MysticalOS** [comment]
I deleted comment cause i implemented hard stop locally instead for the nameplate CD issue.

btw i want to make sure. if using hide with force, that's only removing custom nameplate auras and not CDs right? CDs should still be entirely handled by whether or not a timer is present or not and not ben listening to the force hide call.

**Comment by cont1nuity** [comment]
> btw i want to make sure. if using hide with force, that's only removing custom nameplate auras and not CDs right? CDs should still be entirely handled by whether or not a timer is present or not and not ben listening to the force hide call.

Correct, icon and timer handling is separated.

**Comment by MysticalOS** [comment]
I want to have at least preliminary version in a mergable state for Oct 31st reset, so it gets at least one week of testing prior to patch. If we have to, can split it into two diff PRs but want to give people time to get used to it, configure it etc BEFORE patch goes live so they aren't fumbling with it patch day (on top of 100 other things).

**Comment by cont1nuity** [comment]
`DBM.Options.NPIconGrowthDirection` still needs some option in the menu for it, but apart from that I think we're good for a test run. :)
I have not seen any breaking issues so far while testing locally comparing to Plater behavior directly.

**Comment by MysticalOS** [comment]
was NPIconGrowthDirection only thing you added that needs UI? i thought i remember 4 variables? I'll take a look

**Comment by cont1nuity** [comment]
> > I am thinking about adding some option for direction of the icon row growth. Along with the offset, this would allow nice positioning, e.g. upper left corner, growing right.
> 
> This is implemented now, but option is not added to the menu: `DBM.Options.NPIconGrowthDirection` takes `"UP"`, `"DOWN"`, `"LEFT"`, `"RIGHT"` and `"CENTER"` (default if unset)

This is what I added. The values can be adjusted.
The other options for offset were already added.

**Comment by MysticalOS** [comment]
@QartemisT can you add some pretty UI options in the nameplate section for this? probably take you 2 minutes

**Comment by MysticalOS** [comment]
oh using common local, smart. Might as well add one for Center though
